<a href="https://colab.research.google.com/github/AllanBottino/Series_Temporais_Python/blob/main/Resample_vs_Asfreq_S%C3%A9ries_Temporais_Python_Parte_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Resample() vs Asfreq()

No artigo anterior eu coloquei apenas o uso do "resample()", e não do "asfreq()" neste eu vou tentar explicar quando usar um e quando usar o outro.
Já aviso que essa é uma explicação simples, é possível aprofundar muito mais, pois quando se trata de reamostragem inevitavelmente entraremos em Upsampling e Downsampling, falarei de Upsampling e Downsampling no próximo artigo, mas creio que em termos práticos não me aprofundando e visando o código em si, e uma aplicação simples, o código será útil.
links de onde eu pesquisei principalmente:

https://stackoverflow.com/questions/18060619/difference-between-asfreq-and-resample

https://machinelearningmastery.com/resample-interpolate-time-series-data-python/#:~:text=Resampling%20involves%20changing%20the%20frequency,as%20from%20days%20to%20months.


# Carregando as bibliotecas:

In [2]:
import pandas as pd

import numpy as np

Nesse especifico caso vou usar números aleatórios, pois é mais fácil de se entender com esse exemplo do que com números reais:

In [3]:
ts = pd.Series(range(365), index = pd.date_range(start='20190101',  end='20191231',freq = 'D'))

ts.head()

2019-01-01    0
2019-01-02    1
2019-01-03    2
2019-01-04    3
2019-01-05    4
Freq: D, dtype: int64

Agora reamostraremos os dados acima por trimestre usando ".asfreq("Q")":


In [4]:
ts.asfreq('Q')

2019-03-31     89
2019-06-30    180
2019-09-30    272
2019-12-31    364
Freq: Q-DEC, dtype: int64

Agora vamos ver essa série de tempo por trimestre usando o resample:

In [5]:
ts.resample('Q')

Resample retorna um "DatetimeIndexResampler", ou um "Resample Object":

https://pandas.pydata.org/docs/reference/resampling.html

Não conseguimos ver o que está dentro. Para vermos o que está dentro vamos usar o bins, maiores informações de bins: https://predictivehacks.com/how-to-create-bins-in-python-using-pandas/

***Quando vocês estiverem usando o resample em algo real seria algo do tipo: "ts.resample("Q").sales.mean(), tem que especificar qual coluna você quer ver.

In [6]:
bins = ts.resample('Q')
bins.groups

{Timestamp('2019-03-31 00:00:00', freq='Q-DEC'): 90,
 Timestamp('2019-06-30 00:00:00', freq='Q-DEC'): 181,
 Timestamp('2019-09-30 00:00:00', freq='Q-DEC'): 273,
 Timestamp('2019-12-31 00:00:00', freq='Q-DEC'): 365}

Aconteceu um erro de uma unidade, mas vamos relevar esse erro. Vamos calcular a média de cada trimestre e entenderemos o problema real:

In [7]:
print("Asfreq por si só:\n", ts.asfreq("Q"),"\n")

print("Vendas do dia 2019-03-31:", ts.loc["2019-03-31"])
print("Vendas do dia 2019-06-30:",ts.loc["2019-06-30"])
print("Vendas do dia 2019-09-30:",ts.loc["2019-09-30"])
print("Vendas do dia 2019-12-31:", ts.loc["2019-12-31"],"\n")

print("Asfreq com a MÉDIA: \n", ts.asfreq('Q').mean())

Asfreq por si só:
 2019-03-31     89
2019-06-30    180
2019-09-30    272
2019-12-31    364
Freq: Q-DEC, dtype: int64 

Vendas do dia 2019-03-31: 89
Vendas do dia 2019-06-30: 180
Vendas do dia 2019-09-30: 272
Vendas do dia 2019-12-31: 364 

Asfreq com a MÉDIA: 
 226.25


Agora vejam essa conta:

(89+180+272+364)/4 = 226.25

Foi gerado uma média de todos os valores. Observe que esta não é a média de cada trimestre, mas a média do último dia de cada trimestre.

Para calcular a média de cada trimestre:

In [8]:
print(ts.resample('Q').mean(),"\n")

print("Vendas de 2019-01-01 até 2019-03-31:",ts.loc["2019-01-01" : "2019-03-31"].mean())
print("Vendas de 2019-03-31 até 2019-06-30:",ts.loc["2019-03-31" : "2019-06-30"].mean())
print("Vendas de 2019-06-30 até 2019-09-30:",ts.loc["2019-06-30" : "2019-09-30"].mean())
print("Vendas de 2019-09-30 até 2019-12-31:",ts.loc["2019-09-30" : "2019-12-31"].mean())

2019-03-31     44.5
2019-06-30    135.0
2019-09-30    226.5
2019-12-31    318.5
Freq: Q-DEC, dtype: float64 

Vendas de 2019-01-01 até 2019-03-31: 44.5
Vendas de 2019-03-31 até 2019-06-30: 134.5
Vendas de 2019-06-30 até 2019-09-30: 226.0
Vendas de 2019-09-30 até 2019-12-31: 318.0


A principal diferença entre os dois é que o resample() é fundamentalmente uma agregação de dados, ele faz o mesmo que o groupby mas com um periodo de tempo pré determinado (Offset aliases), por isso podemos fazer a média ou soma ou qualquer outra operação tendo como base o periodo de tempo escolhido (Offset aliases). Enquanto que o asfreq()é fundamentalmente uma seleção de dados, os asfreq acaba que sendo algo mais simples que o resample. Quando usar um ou outro? Aí vai depender da sua necessidade, creio que com esse simples exemplo tenha ficado claro de como usar os comandos e nos futuros usos do resample e do asfreq um cuidado irá ser tomado.

Um artigo curto, mas de fácil entendimento. No próximo artigo falarei de Downsampling e Upsampling, acompanhem, pois, vocês verão que o assunto se conecta com esse.

Abraços.